### PR TITLE
nip96: upload images

### LINKED ISSUE #466: Image Upload in Kind1

### PR DESCRIPTION
**Pull request by bob2402**
resolved #466

### CODE REVIEW

File: app/UI/app_update.tsx [hunk 1]
```diff
@@ -277,10 +280,79 @@ const handle_update_event = async (chan: PutChannel<true>, args: {
...
+            }
+            const api_url = "https://nostr.build/api/v2/nip96/upload";
+            const { uploadedURLs, errorMessages } = await uploadValidFiles(app, valid_files, api_url);
+            if (errorMessages.length > 0) {
+                app.toastInputChan.put(() => errorMessages.join("\n"));
+            } else {
+                app.toastInputChan.put(() => "uploaded");
+            }
+            app.eventBus.emit({
+                "type": "TextAppention",
```
Comment (BlowaterNostr) [inline app/UI/app_update.tsx:353]: remove this event. A UI event is only triggered by direct user interaction. This is just an implementation detail.

File: app/UI/app_update.tsx [hunk 2]
```diff
@@ -277,10 +280,79 @@ const handle_update_event = async (chan: PutChannel<true>, args: {
...
+                return { uploadedURLs, errorMessages };
+            }
+            const { files } = event;
+            if (files.length === 0) {
+                app.toastInputChan.put(() => "no files selected");
+                continue;
+            }
+            const invalid_files_name: string[] = [];
+            const valid_files: File[] = [];
+            files.forEach((file) => {
```
Comment (BlowaterNostr) [inline app/UI/app_update.tsx:327]: Please, don't default to `forEach`. This has been stated many times.

File: app/UI/app_update.tsx [hunk 3]
```diff
@@ -277,10 +280,79 @@ const handle_update_event = async (chan: PutChannel<true>, args: {
...
+            }
+            if (invalid_files_name.length > 0) {
+                app.toastInputChan.put(() =>
+                    `Uploading, except for these invalid files: ${invalid_files_name.join(", ")}`
+                );
+            } else {
+                app.toastInputChan.put(() => "uploading...");
+            }
+            const api_url = "https://nostr.build/api/v2/nip96/upload";
+            const { uploadedURLs, errorMessages } = await uploadValidFiles(app, valid_files, api_url);
```
Comment (BlowaterNostr) [inline app/UI/app_update.tsx:346]: inline `uploadValidFiles`. There is no need to create a local function

File: app/UI/app_update.tsx [hunk 4]
```diff
@@ -277,10 +280,79 @@ const handle_update_event = async (chan: PutChannel<true>, args: {
...
+                const max_byte_size = 10485760;
+                const image_postfix = [".png", ".jpg", ".jpeg", ".gif", ".webp"];
+                if (file.size > max_byte_size) {
+                    return false;
+                }
+                if (!image_postfix.some((postfix) => file.name.endsWith(postfix))) {
+                    return false;
+                }
+                return true;
+            }
```
Comment (BlowaterNostr) [inline app/UI/app_update.tsx:295]: This logic should be in `nostr.ts` if you choose to do so, or we don't need to do it here.

File: app/UI/app_update.tsx [hunk 5]
```diff
@@ -277,10 +280,79 @@ const handle_update_event = async (chan: PutChannel<true>, args: {
...
+            async function uploadValidFiles(app: any, validFiles: File[], apiURL: string): Promise<{
+                uploadedURLs: string[];
+                errorMessages: string[];
+            }> {
+                const uploadedURLs: string[] = [];
+                const errorMessages: string[] = [];
+                const uploadPromises = validFiles.map(async (file) => {
+                    try {
```
Comment (BlowaterNostr) [inline app/UI/app_update.tsx:303]: which part of this function throws?

**Review by BlowaterNostr — CHANGES_REQUESTED** [review]

File: app/UI/app_update.tsx (hunk 1 above)
Comment (bob2402) [reply app/UI/app_update.tsx]: This part is completed using callback function.

File: app/UI/app_update.tsx (hunk 4 above)
Comment (bob2402) [reply app/UI/app_update.tsx]: Okay, we don't need to do it here.

File: app/UI/app_update.tsx (hunk 3 above)
Comment (bob2402) [reply app/UI/app_update.tsx]: Adjusted the code, this part of the suggestion is outdated.

File: app/UI/app_update.tsx (hunk 5 above)
Comment (bob2402) [reply app/UI/app_update.tsx]: Adjusted the code, this part of the suggestion is outdated. Thanks anyway.

File: app/UI/app_update.tsx (hunk 2 above)
Comment (bob2402) [reply app/UI/app_update.tsx]: Thanks, I'll avoid it later.

**Review by BlowaterNostr — APPROVED** [review]

**Pull request merged**